### DISCUSSION TITLE
[Question][Hw1] What does Value.t represent?

### DISCUSSION THREAD

**Discussion by pingpingy1** [prosyslab-classroom/cs424-program-reasoning#388] [Q&A]
(Forgive me for the trivial questions... I am new to LLVM...)

I originally thought that `Variable.t` should represent the name of the variable, collected from the instruction object with the `Llvm.value_name` function.
For example, when the object `instr` encodes the instruction
`%x = add nsw i32 10 20`
then the result of the `add` instruction is saved in the `%x` variable.
I had thought that the `%x` should be used for the Value.t somehow, and that `%x` should be parsed from `Llvm.value_name instr` command.
However, it seems that Variable.t takes the data type of Llvm.llvalue.
Then, how should I interpret what Variable.t means? How is it encoded in the instruction?

Thank you.

**Comment by pingpingy1** [comment]
Reading the LLVM primer again more thoroughly, it says the following:

>  Notice that the first operand of `%inc = add nsw i32 %x, 1` is variable `%x` that also represents the first instruction.

Does this mean that we may use the instruction object as the Variable.t that specifies the variable?

**Comment by KihongHeo** [comment]
Yes. You are right.

Don't say "forgive me". I don't expect anyone in our class is familiar with LLVM. If you have anything unclear, feel free to ask.
